### PR TITLE
Rewrite and centralise generic healthcheck docs

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/zrq6ih4v/528-investigate-removal-of-no-processes-found-for-x-for-apps-with-a-healthcheck

Please see the commits for more details.

### CODE REVIEW

File: source/manual/alerts/whitehall-app-healthcheck-not-ok.html.md [hunk 1]
```diff
@@ -18,8 +20,8 @@ Production.
 Run the `publishing:scheduled:requeue_all_jobs` Rake task to requeue all
 scheduled editions:
 
-- [Integration][]
-- [Staging][]
+- [Run in Integration Jenkins][]
+- [Run in Staging Jenkins][]
```
Comment (1pretz1) [inline source/manual/alerts/whitehall-app-healthcheck-not-ok.html.md:24]: These two links are now broken

**Review by 1pretz1 — COMMENTED** [review]
Nice work consolidating these docs @benthorner, much better!

Another thing I thought of concerning the generic app healthcheck doc, could we add a section in around timeouts? They seem to be a fairly common alert and it'd be good to have some information / steps on what to do if a healthcheck is timing out.

Some ideas for the steps could be:

* checking the app is running (upstart, unicorn etc)
* checking the app is serving requests (kibana, app or nginx logs)
* if so checking for high load / traffic and slow requests

File: source/manual/alerts/app-healthcheck-not-ok.html.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+Many apps on GOV.UK have a healthcheck endpoint.
+
+- This is usually `/healthcheck` [[1](https://github.com/alphagov/govuk-puppet/blob/2693343ebc1aced7a7f94e8aba31fee8b05df8a5/modules/govuk/manifests/apps/email_alert_api.pp#L166)].
+- Some apps just pick a random page [[1](https://github.com/alphagov/govuk-puppet/blob/2693343ebc1aced7a7f94e8aba31fee8b05df8a5/modules/govuk/manifests/apps/collections.pp#L51)].
+
+The alert works by making a request for the healthcheck endpoint on the machine where the app is running. If you need to test the healthcheck endpoint manually, you can SSH on to the machine and `curl` it yourself.
+
+```
+# SSH on to a machine running Content Publisher
+gds govuk connect -e integration backend
```
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/alerts/app-healthcheck-not-ok.html.md]: Oops, thanks for spotting. Fixed all these.

File: source/manual/alerts/app-healthcheck-not-ok.html.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+# Find the port it's running on
+ps -ef | grep content-publisher
```
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/alerts/app-healthcheck-not-ok.html.md]: 👍

File: source/manual/alerts/app-healthcheck-not-ok.html.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+# Do the Icinga check manually
+curl localhost:3221/healthcheck
+```
+
+Apps with a custom healthcheck endpoint often make use of [the generic checks in govuk_app_config](https://github.com/alphagov/govuk_app_config/blob/master/docs/healthchecks.md). Some apps also implement custom checks, and the alert should link to custom documentation to explain these.
```
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/alerts/app-healthcheck-not-ok.html.md]: Sure, I've gone with one in Content Publisher.

File: source/manual/alerts/app-healthcheck-not-ok.html.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+## ActiveRecord
+
+This means the app is unable to connect to its database.
+
+- Check for [any RDS alerts](/manual/govuk-in-aws.html#postgresql-and-mysql).
+- Check the [AWS RDS dashboard][postgres dash] to see if we're experiencing resourcing issues.
+- Check for network connectivity to the DB.
+
+```
+# SSH on to machine with problem
```
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/alerts/app-healthcheck-not-ok.html.md]: You no like terse English? Updated.

File: source/manual/alerts/whitehall-app-healthcheck-not-ok.html.md (hunk 1 above)
Comment (benthorner) [reply source/manual/alerts/whitehall-app-healthcheck-not-ok.html.md]: Oops, fixed!

**Comment by benthorner** [comment]
Thanks @1pretz1! I've addressed all the individual comments. Good point about the timeout: I've added a couple of sections at the top for the two kinds of error that you might see.

Could you take another look?

File: source/manual/alerts/app-healthcheck-not-ok.html.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+## Connection Refused Error
+
+This means the app is not accepting requests for the healthcheck endpoint, and is probably down.
+
+- Check [the processes are running](check-process-running.html).
+- Try the healthcheck endpoint manually, as above.
+
+## Timeout Error
+
+This means the app is responding to requests, but the request is taking too long (over [20 seconds](https://github.com/alphagov/govuk-puppet/blob/fab936cb82be7fad42636fcafca3718a8368ebfe/modules/icinga/files/usr/lib/nagios/plugins/check_json_healthcheck#L155)).
```
Comment (1pretz1) [inline source/manual/alerts/app-healthcheck-not-ok.html.md:40]: > This means the app is responding to requests

I don't think this is necessarily true. In the recent incident where email-alert-api was failing to start the healthcheck was timing out even though the app wasn't (couldn't) respond to requests. I believe the healthcheck requests made it into Nginx but then timed out waiting to communicate with the app.

Maybe just remove that statement ?

**Comment by 1pretz1** [comment]
Looks good, just one minor comment that needs addressing before 🚢

File: source/manual/alerts/app-healthcheck-not-ok.html.md (hunk 6 above)
Comment (benthorner) [reply source/manual/alerts/app-healthcheck-not-ok.html.md]: That's true. I've amended to say "accepting" requests, but not necessarily responding: the unicorn worker starts and opens a connection, but then encounters a fatal error and exits (hence the timeout). 

I've also added an extra "check the logs" step to cover this.

**Review by 1pretz1 — APPROVED** [review]
💯

**Pull request merged**